Refine multi-step PR instructions #27

damithc opened the pull request:
Fixes #27 

Overview:
```
We use multi-step format for big PRs only. 
Having two workflows is confusing to reviewers. 

Update PR workflow for all PRs to follow one format.
The new format requires in every PR,
* proper commit organization
* use of CanIHasReview for requesting reviews
```
Do after merging this PR:
- [ ] Create an issue for adding an example of a 'bad commit organization'

damithc commented:
@pyokagan review when you can. No hurry.

damithc commented:
@pyokagan shall I update this PR based on our slack discussion? That is, move this description of multi-step PRs to [this doc](https://github.com/oss-generic/process/blob/master/docs/QuestionsIssuesPrs.md#submitting-a-pr). Also, there's no need to  use  the term 'multi-step' any more because that's the only way?

pyokagan commented:
@damithc 
> @pyokagan shall I update this PR based on our slack discussion?

Sure, please go ahead. Sorry I couldn't get to reviewing this since I haven't had a large enough block of time yet.

damithc commented:
> Sorry I couldn't get to reviewing this since I haven't had a large enough block of time yet.

No worries. Not urgent. Refining workflow cannot be rushed :-p

pyokagan commented on docs/HowToGuides.md:
```diff
@@ -2,76 +2,19 @@
 
     <details>
-    <summary>Insructions for the reviewer</summary>
+    <summary>Insrtuctions for the reviewer</summary>
```
Close, but not quite :-P

pyokagan commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -39,28 +40,62 @@ license as that used by the project.
+     - [x] Each commit contains a single logical change, and this change must stand on its own. 
+       i.e. each commit has a single responsibility, and this responsibility must be fully carried out.<br>
+       For example, if the commit message says `Move delete() from Person class to Address class`, the commit cannot 
+       contain the addition of `delete()` to `Address` class only; it should also contain the deletion of `foo()` from
```
Did you mean `foo()` -> `delete()`?

pyokagan commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -103,7 +152,38 @@ license as that used by the project.
+   <summary>Guidelines for updating the PR in response to a review</summary>
+   
+   * Updates to existing logical changes should be done by 
+     [modifying their corresponding commit](http://sethrobertson.github.io/GitPostProduction/gpp.html#analyze).
```
I fully read through http://sethrobertson.github.io/GitPostProduction/gpp.html#analyze this time, and I think it's not a very good explanation of how to edit a commit since it focuses more on ensuring that all commits in a PR passes tests rather than explaining how to edit a *specific* commit.

The best I could find via a quick search is http://stackoverflow.com/a/1186549

pyokagan commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -20,7 +20,8 @@ If you don't get an answer within 1-2 days, feel free to post another comment in
 
-> This is an adaptation of the [_GitHub Flow_](https://guides.github.com/introduction/flow/). 
+> _Ask a programmer to review 10 lines of code, he'll find 10 issues. Ask him to do 500 lines and he'll say it 
+> looks good. [[source](https://twitter.com/girayozil/status/306836785739210752)]_
```
Hrm it seems kind of strange to have this quote directly below the "Submitting a PR" heading.

Maybe we can remove this quote for now? If we ever had a document or section dedicated specifically to good commit organization, we could add this quote back in again.

damithc commented:
Sorry @pyokagan, I was halfway through the edits and pushed to see how it renders and also to update branch so that I can continue later using office computer. So, the doc was not in a review-ready state. :-p
Anyway, the comment are still useful. I'll incorporate them and let you know when ready for review again.

CanIHasReview commented:
# v1

@damithc submitted v1 for review.

* [v1 1/3] [Document 'commit organization' requirements](https://github.com/oss-generic/process/pull/29/commits/afadb0a44743facc310cad8f37aa8c8790b06eae)
* [v1 2/3] [Require proper commit organization in PRs](https://github.com/oss-generic/process/pull/29/commits/bbbb18ddcaa7a103e3f97fa61e19c2200f5b867a)
* [v1 3/3] [Update instructions for reviewing PRs](https://github.com/oss-generic/process/pull/29/commits/dd5a79636502dd6966a4f4e0189e9291ecfd208b)

([:books: Archive](https://github.com/oss-generic/process/compare/da576cbd413c184c4571d9969da4489e4b9d18d0...dd5a79636502dd6966a4f4e0189e9291ecfd208b))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/29/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

CanIHasReview commented:
# v2

@damithc submitted v2 for review.

* [v2 1/3] [Document 'commit organization' requirements](https://github.com/oss-generic/process/pull/29/commits/9420a3a31eef630a6a8974f43b0fdab068e7c022)
* [v2 2/3] [Require proper commit organization in PRs](https://github.com/oss-generic/process/pull/29/commits/7f0272bc4dc8a65b708267e46c75d950e9cc6bf0)
* [v2 3/3] [Update instructions for reviewing PRs](https://github.com/oss-generic/process/pull/29/commits/bafba4435628bfac37287927f1bba7113565ab1b)

([:books: Archive](https://github.com/oss-generic/process/compare/da576cbd413c184c4571d9969da4489e4b9d18d0...bafba4435628bfac37287927f1bba7113565ab1b)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/oss-generic/process/commit/f309571c668f7deff7bfe8ca1851c21557346b5d))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/29/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented:
@pyokagan updated based on comments.

pyokagan commented:
@damithc So what do you think about removing the "Additional Requirements" distinction? Instead, we could have a "tips for `d.FirstTimer` issues" instead, where we recommend that contributors work with only one commit instead (i.e. use `git commit --amend` only.)

damithc commented:
> So what do you think about removing the "Additional Requirements" distinction?

Will respond in slack.

damithc commented:
@pyokagan One possible option is to revise as you have suggested but instruct reviewers to be lenient on first time contributors. What do you think?

pyokagan commented:
@damithc 
> instruct reviewers to be lenient on first time contributors.

Could you give some examples? In what situations would it be better for reviewers to be lenient, as opposed to explaining to the first time contributors what they have done wrong and what they could do to fix the problem (ideally with the help of pre-existing documentation?)

Some common mistakes that I guess first time contributors will make:

* Merging to sync the branch with master rather than rebasing

    * CanIHasReview already automatically rejects PRs with merge commits. Reviewers can prompt PR authors to perform a rebase (which will automatically drop merge commits)

* PR branch is outdated

    * CanIHasReview already automatically rejects out-of-date branches. Reviewers can prompt PR authors to perform a rebase.

* Having bad commit messages

    * Reviewers can write the commit messages for the PR author.

* PR has many junk commits

    * Reviewers can tell the contributor to squash all the commits together (ideally, we will have a document which we will direct them to)

pyokagan commented:
We may actually wish to break down the "first time contributor" group into two groups of people:

* People who are making their first contribution with the intent of contributing further in the near future (e.g. for GSoC)

    * We should ensure that these contributors follow our full workflow

* People who are making a "drive-by" contribution without the intent of contributing further

    * We can help them clean up and merge in their contribution for them (e.g. by helping them to clean up any unrelated changes in their PR and merge in the result under their name)

damithc commented:
@pyokagan 
The examples of common mistakes you have given are accurate and the suggested corrective actions are good. No, we should not accept bad code (we don't accept bad code even now). As you suggested, the only leniency we can show is to write the commit message ourselves if needed. Shall I update the docs along those lines?

pyokagan commented:
@damithc 
> As you suggested, the only leniency we can show is to write the commit message ourselves if needed

Yep. I think the rest of the Git stuff (e.g. how to rebase and squash) is best solved by having better documentation, rather than trying to fallback on things like merge commits.

It makes sense to be more lenient about commit messages though, as there is no clear "best" way to writing them, plus we need to consider that some contributors are not native English speakers as well. It should be overall less effort for us if we just write the commit messages for them. They would still need to actually do the action of amending the commit messages, though, unless we take advantage of GitHub's feature of [allowing maintainers to push to PR branches](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)

> Shall I update the docs along those lines?

I'm OK with this.

CanIHasReview commented:
# v3

@damithc submitted v3 for review.

* [v3 1/3] [Document 'commit organization' requirements](https://github.com/oss-generic/process/pull/29/commits/99a9d06f5f4e48b08d842c1269f98dc55727ec48)
* [v3 2/3] [Require proper commit organization in PRs](https://github.com/oss-generic/process/pull/29/commits/42271e6477e4b445c80cc29b3dd65e010314eb9b)
* [v3 3/3] [Update instructions for reviewing PRs](https://github.com/oss-generic/process/pull/29/commits/0413dde4ac67b5d85bab544ffdfe4c788af36de9)

([:books: Archive](https://github.com/oss-generic/process/compare/da576cbd413c184c4571d9969da4489e4b9d18d0...0413dde4ac67b5d85bab544ffdfe4c788af36de9)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/oss-generic/process/commit/8c1ef264b5b35cd047fd2c0f5117411befdaca32))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/29/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented on docs/FormatsAndConventions.md:
```diff
@@ -41,6 +41,56 @@ Here is an example (adapted from [this commit](https://github.com/CS2103AUG2016-
+  > * [Our conventions for commit messages](#commit-message)
+  > * [Web article] _[How to Write a Git Commit Message](http://chris.beams.io/posts/git-commit/)_ 
+
+<p>
```
Without `<p>` the two block quotes get stuck together no matter how many blank lines are between them.

damithc commented:
@pyokagan 

* Removed 'additional requirements' distinction
* Added an `overview` section to the PR description, to be used as body of merge commit.

damithc commented:
@pyokagan 
1. Should we discourage commenting in the interdiff? Those comments don't appear in the comment thread of the PR.
2. A PR branch can get outdated after the commit summary has been posted. In that case we should review the outdated version first and delay rebasing till next iteration?
3. Should we require a PR overview (like given in this PR) for PRs having multiple commits, to be used as the body of the merge commit?

pyokagan commented:
@damithc 
> Should we discourage commenting in the interdiff?

Yeah, we shouldn't be commenting in the interdiff. The interdiff is just meant to be used as a guide, reviewers should still look at the changes in the context of its associated commit.

With that said, I don't think we need to mention this in the workflow document -- this is just a side effect of how CanIHasReview currently implements interdiffs. I plan to make CanIHasReview link to GitHub's compare page instead, rather than the commit view page, so that we won't accidentally add comments to it. e.g. linking to https://github.com/oss-generic/process/compare/8c1ef264b5b35cd047fd2c0f5117411befdaca32~1...8c1ef264b5b35cd047fd2c0f5117411befdaca32

> A PR branch can get outdated after the commit summary has been posted. In that case we should review the outdated version first and delay rebasing till next iteration?

Yes. A rebase is counted as "modifying" the PR, so a new version should be submitted if the PR is rebased.

> Should we require a PR overview (like given in this PR) for PRs having multiple commits, to be used as the body of the merge commit?

Yes. In fact, I think it would be good to have a PR overview for all PRs actually, not just those with more than one commit, since we would need to have something to put in the merge commit anyway.

But before putting that in the workflow, I think we should resolve #32 first.

damithc commented:
> But before putting that in the workflow, I think we should resolve #32 first.

Yes, let's wait to add that.

pyokagan commented:
@damithc I guess you can make the changes to the PR first :-)

damithc commented:
@pyokagan 

> I guess you can make the changes to the PR first

Did you review v3?  Or shall I do v4 now?

pyokagan commented:
@damithc 
> Did you review v3? Or shall I do v4 now?

I didn't review v3, just didn't want to block you since we have already agreed (?) on some changes in https://github.com/oss-generic/process/pull/29#issuecomment-275902056. I'll review both v3 and v4 together.

damithc commented:
> I'll review both v3 and v4 together

Roger that.

CanIHasReview commented:
# v4

@damithc submitted v4 for review.

* [v4 1/3] [Document 'commit organization' requirements](https://github.com/oss-generic/process/pull/29/commits/99a9d06f5f4e48b08d842c1269f98dc55727ec48)
* [v4 2/3] [Require proper commit organization in PRs](https://github.com/oss-generic/process/pull/29/commits/42271e6477e4b445c80cc29b3dd65e010314eb9b)
* [v4 3/3] [Update instructions for reviewing PRs](https://github.com/oss-generic/process/pull/29/commits/c44c2dfe24e88e2d75d8f6cb21d04b1e4d34531e)

([:books: Archive](https://github.com/oss-generic/process/compare/da576cbd413c184c4571d9969da4489e4b9d18d0...c44c2dfe24e88e2d75d8f6cb21d04b1e4d34531e)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/oss-generic/process/commit/59a0d9f9ffed4d4a0b103cbbfd568c69da81b107))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/29/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

pyokagan commented on docs/FormatsAndConventions.md:
```diff
@@ -41,6 +41,56 @@ Here is an example (adapted from [this commit](https://github.com/CS2103AUG2016-
+  
+  > Rationale: Bigger commits make reviewing harder.
+
+  Commits containing _**automated changes**_ (e.g. changes resulting from automated refactorings),
```
Maybe we should call this a "mechanical change" instead? E.g. Cut-and-paste wouldn't be considered an "automated change", right? But I would consider that code movement, which would belong under this category.

pyokagan commented on docs/FormatsAndConventions.md:
```diff
@@ -41,6 +41,56 @@ Here is an example (adapted from [this commit](https://github.com/CS2103AUG2016-
+
+> **Note for first time contributors**:
+> 
+> * PRs for `d.FirstTimers` issues are usually simple enough to be contained in one commit. 
```
Maybe we should add this requirement to the DefininingLabels document as well, e.g. `d.FirstTimers issues MUST be simple enough to be contained in one commit.`

pyokagan commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -20,95 +20,166 @@ If you don't get an answer within 1-2 days, feel free to post another comment in
+   Create a branch off the `master` branch. 
+   Follow [our naming conventions for branch names](FormatsAndConventions.md#branch).
+   
+   > Common mistake: Branch off from the currently active branch instead of the `master` branch.
```
I wonder if we should phrase this as a gigantic `WARNING` instead, e.g.

```
WARNING: Make sure you branch off master rather than the currently active branch.
Otherwise, you may get commits from other branches inside your PR.
```

CanIHasReview commented:
# v5

@damithc submitted v5 for review.

* [v5 1/3] [Document 'commit organization' requirements](https://github.com/oss-generic/process/pull/29/commits/a63eec82d83f17c6c11bf76c180e480e8f4d4b24)
* [v5 2/3] [Require proper commit organization in PRs](https://github.com/oss-generic/process/pull/29/commits/fcb278bb507b6307febad8cd08c1d6624444decf)
* [v5 3/3] [Update instructions for reviewing PRs](https://github.com/oss-generic/process/pull/29/commits/b2f9f4178d52819a04d9fc8d178eedb89ccbbd3c)

([:books: Archive](https://github.com/oss-generic/process/compare/28635c4c7336f8b04d1546626ad75c76c27eb736...b2f9f4178d52819a04d9fc8d178eedb89ccbbd3c)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/oss-generic/process/commit/d01930edaa726bf81f2e60480a0fa65ea14604ab))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/29/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented:
@pyokagan updated as suggested.

pyokagan commented:
@damithc I'll let you do the merges for oss-generic :-P

damithc commented:
> I'll let you do the merges for oss-generic :-P

Roger that :-)
